PR: Revise docstrings in the `spyder.api.plugins` module

CAM-Gerlach opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] ~Added unit test(s) covering the changes (if testable)~
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] ~Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))~


<!--- Explain what you've done and why --->

Required by spyder-ide/spyder-api-docs#18

* Thoroughly revise the docstring content for the `spyder.api.plugins` modules:
	* Ensure all modules and APIs have a proper summary line
	* Document un(der) documented public APIs
	* Greatly clarify and expand existing descriptions
	* Add all param, return and attribute types to both signatures and docstrings
	* Fix and clarify many instances of missing, incorrect or unclear param types in the signature and docstrings related to parameters accepting `None` and/or having a default value, and clarify in the docstring what that value is and its meaning.
	* Add and fix Sphinx roles and directives throughout
	* Fix all Sphinx broken references in the affected modules



### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: CAM-Gerlach

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
Spyder is crashing with these changes. The failure was introduced in commit 2d7c2ecb6fe838a61d52e54203cb480a5dc1bdce.

CAM-Gerlach commented:
Hmm, seems my smoke tests locally didn't catch it nor did my searches for usages of `PLUGIN_NAME` from the plugin instance rather than the widget/container, but reproing it locally and digging deeper I see that `PLUGIN_NAME` on the plugin instance is `getattr`ed and setattred on `SpyderWidgetMixin` in [SpyderWidgetMixin.__init__](https://github.com/spyder-ide/spyder/blob/5e701222c5f16069924f15808b75c29d45e7b2df/spyder/api/widgets/mixins.py#L679), whichdidn't play nice with it being a `@property`, so I reverted that part of the change (and also found and fixed a missing docstring/type annotation with the return type of the `main` property setter).

In addition, while initially checking for possible other sources of test issues I also noticed that the `configuration` arg to `SpyderPluginV2` originally _did_ have a default value, of `None`, which somehow got inadvertently dropped in my commits (leading me to think it wasn't there originally), and as you suspected it was indeed relied on for several tests. I restored it, as well as updated the relevant type hints in the signatures and docstrings and organized the private attribute setting block in the `__init__` a bit. Between these two, this should fix all the errors/failures in the tests.

ccordoba12 commented on spyder/api/plugins/enum.py:
```diff
@@ -1,69 +1,160 @@
+    """
+
     EnvManager = "spyder_env_manager"
+    """The Spyder Env Manager plugin."""
```
```suggestion
    """The Spyder Environment Manager plugin."""
```

We can be more explicit about this plugin.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+
+    If a plugin uses a widget from another plugin, that other plugin must
+    be declared as a required dependency.
+
```
```suggestion
```

This is not right, e.g. Projects uses widgets from Explorer but it doesn't declare it in its REQUIRES (I guess that was the intention when this API was originally designed, but we didn't follow that direction).

CAM-Gerlach replied:
Okay sure, that was just what it said originally.

CAM-Gerlach replied:
Applied as a new suggestion with correct whitespace handling

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+        dict[str, list[type[SpyderPreferencesTab]]] | None
+    ) = None
+    """
+    Define additional tabs of option for other plugins' configuration pages.
```
```suggestion
    Define additional tabs of options for other plugins' configuration pages.
```

Seems missing.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
     """
-    This signal can be emitted to request the main application to garbage
-    collect deleted objects.
+    Request the main application garbage collect deleted objects.
```
This and all other signals below are not shown to be Qt signals in the rendered docs, so the docstrings for them don't make sense in the way you rewrote them.

So, please either add typing to these signals or mention that they are signals in their docstrings.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
     """
-    This signal can be emitted to request the main application to redirect
-    standard output/error when using Open/Save/Browse dialogs within widgets.
+    Request redirection of ``stdout``/``stderr`` when using Open/Save/Browse.
```
```suggestion
    Request redirection of ``stdout``/``stderr`` when using Open/Save/Browse
    dialogs in widgets.
```

The current docstring makes no sense (what `Open/Save/Browse` means?)

CAM-Gerlach replied:
All common docstring standards and conventions (e.g. PEP 257, Numpydoc, Googledoc) require the summary line to fit in one line (and this is relied upon by Sphinx-Autosummary to populate the summary tables, etc).

However, I can do a much better job of shortening this and then adding another line after for more context.

```suggestion
    Request the main app redirect standard out/error within file pickers.
    
    This will redirect :data:`~sys.stdout` and :data:`~sys.stderr` when using
    :guilabel:`Open`, :guilabel:`Save`, and :guilabel:`Browse` dialogs
    within a plugin's widgets.
```

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
-    This signal can be emitted to report an exception from any plugin.
+    Report an exception from a plugin.
+
+    This signal is automatically connected to the main container/widget.
```
```suggestion
```

This is an implementation detail that is not relevant.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -236,119 +406,153 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderConfigurationObserver,
+        .. important::
+
+            Plugins are initialized automatically by Spyder; plugins shouldn't
+            call or override this method directly.
```
```suggestion
            Plugins are initialized automatically by Spyder, so they shouldn't
            call or override this method directly.
```

Make this sentence flow better.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -236,119 +406,153 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderConfigurationObserver,
-        # registries
-        self.PLUGIN_NAME = self.NAME
+        self._widget: PluginMainContainer | None = None
+        self._container: PluginMainContainer | None = None
```
These can be a `PluginMainWidget` too.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -410,43 +614,98 @@ def _unregister(self):
         return self._plugin_path
 
-    def get_container(self):
+    def get_container(self) -> PluginMainContainer | None:
```
This returns `PluginMainWidget` for dockable plugins, but I don't know if it's worth it to document that.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -491,76 +781,110 @@ def get_dockable_plugins(self):
+            If ``True``, all objects that observe all changes on the
+            configuration ``section`` as well as objects that observe
+            partial tuple paths are notified. For example, if the
+            ``option`` ``"opt"`` of ``section`` ``"sec"`` changes, then
```
```suggestion
            ``option`` `opt" of ``section`` "sec" changes, then
```

I find this is styling confusing in the rendered docs, so I think it's better to remove it:

<img width="674" height="83" alt="image" src="https://github.com/user-attachments/assets/435b1e1b-bfef-47a7-968d-0946810b1afb" />

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -572,204 +896,310 @@ def set_conf(
+        Perform actions required before starting a longer-running process.
+
+        Shows a message in the main window's status bar,
+        changes the mouse pointer to a ``WaitCursor``, and starts a spinner.
```
```suggestion
        changes the mouse pointer to a wait cursor, and starts a spinner.
```

Since you removed the reference to Qt, I don't think there's a need to show this as single word and styled as code.

CAM-Gerlach replied:
Yeah, the issue was I couldn't get the xref to work despite trying a lot of different forms.

I'm not sure "wait cursor" is a super commonly-understood term, but I don't know of any better one.

CAM-Gerlach replied:
```suggestion
        and changes the mouse pointer to a wait cursor.
```

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -572,204 +896,310 @@ def set_conf(
+        Perform actions required after starting a longer-running process.
+
+        Clears the message in the Spyder main window's status bar,
+        restores the mouse pointer to the OS default, and stops spinner.
```
```suggestion
        Clears the message in the Spyder main window's status bar
        and restores the mouse pointer to the OS default.
```

This is not correct (and it's not mentioned in the current docstring).

CAM-Gerlach replied:
Yeah I didn't see it in the proximate code here, it just didn't make sense to me why it wouldn't parallel its `before_` variant.

In fact, I don't see any different code in the `before_` callback that would start the spinner, so should it not be mentioned there either? Or what's the deal here?

CAM-Gerlach replied:
Yup, turns out it should be removed from the above too.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -572,204 +896,310 @@ def set_conf(
-        This is useful to set the color scheme of all instances of
-        CodeEditor used by the plugin.
+            used in the Editor, or ``None`` if the Spyder configuration is
+            not available (usually only ever the case in tests).
```
```suggestion
            used in the Editor, or ``None`` if the Spyder configuration
            object is not available (usually only ever the case in tests).
```

Seems missing.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+        Parameters
+        ----------
+        cancelable: bool, optional
+            ``True`` if the close operation can potentially canceled;
```
```suggestion
            ``True`` if the close operation can be potentially canceled;
```

Seems missing.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+    def update_font(self) -> None:
         """
-        This must be reimplemented by plugins that need to adjust their fonts.
+        Reimplemented by plugins that need to adjust their fonts.
```
```suggestion
        This mus be reimplemented by plugins that need to adjust their fonts.
```

It's hard to grasp what this docstring means with the change you made.

CAM-Gerlach replied:
Actually, I think I meant to go back and add a proper summary line to these after the initial pass, but neglected to do so.

Adding that as well as restoring this to how it was

```suggestion
        Modify the font used in the plugin's interface.
        
        This must be reimplemented by plugins that need to adjust their fonts.
```

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+    def update_style(self) -> None:
         """
-        This must be reimplemented by plugins that need to adjust their style.
+        Reimplemented by plugins that need to adjust their style.
```
```suggestion
        This must be reimplemented by plugins that need to adjust their style.
```

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
-        on_option_update that need to be connected before applying those
-        options to our config system.
+        This must be reimplemented by plugins whose containers emit signals
+        that need to be connected.
```
```suggestion
        This must be reimplemented by plugins whose containers emit signals
        that need to be connected before applying options to Spyder's config
        system.
```

The docstring doesn't make sense as it is.

CAM-Gerlach replied:
Oops, I forgot to apply your suggestion from Discord on this

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+
+    TABIFY: list[str] = []
+    """
+    Define a list of plugins next to which we want to to tabify this plugin.
```
```suggestion
    Define a list of plugins next to which we want to tabify this plugin.
```

Remove repeated `to`.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+
+    RAISE_AND_FOCUS: bool = False
+    """
+    Give the plugin focus on switch to plugin calls.
```
```suggestion
    Raise and give the plugin focus on switch to plugin calls.
```

That's missing.

CAM-Gerlach replied:
We discussed it and decided to rename this `FOCUS_WHEN_SWITCHING` in Spyder 6.2.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+    """
+    Give the plugin focus on switch to plugin calls.
+
+    If ``True``, the plugin will be given focus when switched to.
```
```suggestion
    If ``True``, the plugin will be raised and given focus when switched to.
```

CAM-Gerlach replied:
Same as above, this does not appear to be correct according to either the code nor testing.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
     sig_focus_changed = Signal()
     """
-    This signal is emitted to inform the focus of this plugin has changed.
+    Report the focus state of this plugin has changed.
```
Docstrings for all these signals have the same problems as the ones for `SpyderPluginV2`.

CAM-Gerlach replied:
Adding the type documentation here too

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
-    Notes
-    -----
-    This is automatically connected to main container/widget at plugin's
+    This is automatically connected to main container/widget at plugin
```
```suggestion
```

This is an implementation detail that doesn't need to be mentioned in the docstring.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+    ----------
+    plugin: SpyderDockablePlugin
+        The plugin object to show.
+    force_focus: bool | None
```
```suggestion
    force_focus: bool
```

The signature of the signal doesn't allow for this to be `None`.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+        .. important::
+
+            Plugins are initialized automatically by Spyder; plugins shouldn't
+            call or override this method directly.
```
```suggestion
            Plugins are initialized automatically by Spyder, so they shouldn't
            call or override this method directly.
```

Same suggestion as for SpyderPluginV2.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -572,204 +896,310 @@ def set_conf(
         """
-        Show a message in main window's status bar and change the mouse
-        pointer to Qt.WaitCursor when starting a long process.
+        Perform actions required before starting a longer-running process.
```
```suggestion
        Perform actions required before starting a long-running process.
```

`longer` than what? That makes no sense to me.

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1092,327 +1664,557 @@ def __init__(self, parent, configuration):
+            or hiding it with this method. If ``None``, only give it focus when
+            showing, not hiding (i.e. setting ``state`` to ``True``),
+            and only if :attr:`RAISE_AND_FOCUS` is ``True``. If ``False``,
+            the default, don't give it focus regardless.
```
This docstring has the same problem as the one for its corresponding signal (see above).

CAM-Gerlach replied:
Accordingly:

```suggestion
            If ``True``, always give the plugin's widget focus when showing
            or hiding it with this method. If ``False``, the default,
            don't give it focus.
```

ccordoba12 commented on spyder/api/plugins/new_api.py:
```diff
@@ -1092,327 +1664,557 @@ def __init__(self, parent, configuration):
-    def close_window(self, save_undocked=False):
+    def close_window(self, save_undocked: bool = False) -> None:
+        """
+        Close the plugin's window, optionally saving the undocked state.
```
```suggestion
        Close the plugin's undocked window, optionally saving its state.
```

To go with my suggestion for `create_window`.

CAM-Gerlach replied:
Only potential concern here is being specific about _what_ state it is saving (whether it is undocked/docked, its undocked position, size, etc., other state aspects?)

ccordoba12 reviewed (commented):
Thanks @CAM-Gerlach for your work on this!

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    :guilabel:`Console` or another source, but it does not depend on either
+    of those plugins to work. Methods in the plugin that make use of optional
+    plugins must first check that those plugins are available before calling
+    those methods or using their signals and slots.
```
Okay; should we say something about using the decorator, then?

CAM-Gerlach replied:
We decided not to mention it for now.

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    We thus ask plugins to declare whether or not they need
+    web widgets to make it easier to distribute Spyder without them.
+
+    `More information <https://github.com/spyder-ide/spyder/pull/22196#issuecomment-2189377043>`__
```
```suggestion
    See `Spyder issue #22196 <https://github.com/spyder-ide/spyder/pull/22196#issuecomment-2189377043>`__
    for more information.
```

Give this an actual link title.

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+
+    .. code-block:: python
+
+        TABIFY = ["Plugins.Editor"]
```
(Error was in the original, FYI)

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+        or hiding it with this method. If ``None``, only give it focus when
+        showing, not hiding (i.e. setting ``state`` to ``True``),
+        and only if :attr:`RAISE_AND_FOCUS` is ``True``. If ``False``,
+        the default, don't give it focus regardless.
```
I'm just describing the current behavior of the code; in theory, it could perhaps be used to direct keyboard shortcuts to one plugin while not having to show it/viewing another.

As I'd previously mentioned on Discord, I'll make that change in the 6.2 PR as it is out of scope for this doc one as it changes functional behavior (intentional or not), unlike the other changes just removing already broken/private/dead/non-functional code.

The change to eliminate `None` as an allowed value here:

```suggestion
        If ``True``, always give the plugin's widget focus when showing
        or hiding it with this method. If ``False``, the default,
        don't give it focus regardless.
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1092,327 +1664,557 @@ def __init__(self, parent, configuration):
         """
-        Show a message in main window's status bar, change the mouse pointer
-        to Qt.WaitCursor and start spinner when starting a long process.
+        Perform actions required before starting a longer-running process.
```
Same comment as above explaining the meaning, but fine to make the change.

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    Emitted once the plugin is initialized.
     """
 
     sig_quit_requested = Signal()
```
```suggestion
    sig_quit_requested: Signal = Signal()
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    Request that the main Spyder application quit.
     """
 
     sig_restart_requested = Signal()
```
```suggestion
    sig_restart_requested: Signal = Signal()
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    Request that the main Spyder application perform a restart.
     """
 
     sig_status_message_requested = Signal(str, int)
```
```suggestion
    sig_status_message_requested: Signal = Signal(str, int)
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+        redirection.
     """
 
     sig_exception_occurred = Signal(dict)
```
```suggestion
    sig_exception_occurred: Signal = Signal(dict)
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -236,119 +406,153 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderConfigurationObserver,
-    To be used by plugins tracking main window size changes.
     """
 
     sig_mainwindow_moved = Signal("QMoveEvent")
```
```suggestion
    sig_mainwindow_moved: Signal = Signal("QMoveEvent")
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -236,119 +406,153 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderConfigurationObserver,
+        Unmaximize current plugin only if it is not ``plugin_instance``.
     """
 
     sig_mainwindow_state_changed = Signal(object)
```
```suggestion
    sig_mainwindow_state_changed: Signal = Signal(object)
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -236,119 +406,153 @@ class SpyderPluginV2(QObject, SpyderActionMixin, SpyderConfigurationObserver,
+        The new main window state.
     """
 
     sig_focused_plugin_changed = Signal(object)
```
```suggestion
    sig_focused_plugin_changed: Signal = Signal(object)
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
 
     # ---- API: Available signals
     # -------------------------------------------------------------------------
     sig_focus_changed = Signal()
```
```suggestion
    sig_focus_changed: Signal = Signal()
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+    Report the focus state of this plugin has changed.
     """
 
     sig_toggle_view_changed = Signal(bool)
```
```suggestion
    sig_toggle_view_changed: Signal = Signal(bool)
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -781,302 +1211,444 @@ def get_command_line_options(self):
+        the default, don't give it focus regardless.
     """
 
     sig_update_ancestor_requested = Signal()
```
```suggestion
    sig_update_ancestor_requested: Signal = Signal()
```

CAM-Gerlach reviewed (commented):
Add the type hints for the signals as requested

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1,228 +1,398 @@
+    If a plugin uses a widget from another plugin, that other plugin must
+    be declared as a required dependency.
+
+    .. note::
```
```suggestion
    List of names of required dependencies for this plugin.

    .. note::
```

Remove whitespace correctly.

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -410,43 +614,98 @@ def _unregister(self):
+        -------
+        str
+            The path to the directory in which the plugin's ':file:`plugin.py`
+            module is located, i.e. its top-level site package directory.
```
```suggestion
            The path to the directory containing the module with the plugin
            class (:class:`SpyderPluginV2` or :class:`SpyderDockablePlugin`).
```

CAM-Gerlach commented on spyder/api/plugins/new_api.py:
```diff
@@ -1092,327 +1664,557 @@ def __init__(self, parent, configuration):
+        ----------
+        save_undocked : bool, optional
+            ``True`` if the undocked state should be saved. If ``False``,
+            the default, don't persist the undocked state.
```
```suggestion
            ``True`` if the window state (size and position) should be saved.
            If ``False``, the default, don't persist the window state.
```

CAM-Gerlach commented:
@meeseeksdev please backport to 6.x
